File: backend/app/models/game.py
from enum import Enum
from typing import List, Dict, Optional
from pydantic import BaseModel, Field

# --- ENUMS ---
class TrainColor(str, Enum):
    RED = "red"
    BLUE = "blue"
    GREEN = "green"
    YELLOW = "yellow"
    BLACK = "black"
    WHITE = "white"
    ORANGE = "orange"
    PINK = "pink"
    WILD = "wild" # Locomotive

class RouteColor(str, Enum):
    RED = "red"
    BLUE = "blue"
    GREEN = "green"
    YELLOW = "yellow"
    BLACK = "black"
    WHITE = "white"
    ORANGE = "orange"
    PINK = "pink"
    ANY = "any" # Grey routes

class GameStatus(str, Enum):
    LOBBY = "lobby"
    SETUP = "setup"
    ACTIVE = "active"
    FINAL_ROUND = "final_round"
    FINISHED = "finished"

# --- BASE ENTITIES ---
class City(BaseModel):
    id: str
    name: str
    x: float # 0-100 percentage
    y: float # 0-100 percentage

class Route(BaseModel):
    id: str
    node_a: str
    node_b: str
    length: int
    color: RouteColor
    required_wildcards: int = Field(default=0, ge=0)
    owner_id: Optional[str] = None
    sibling_id: Optional[str] = None # For parallel tracks

class TrainCard(BaseModel):
    id: str
    color: TrainColor

class GoalCard(BaseModel):
    id: str
    node_a: str
    node_b: str
    points: int
    is_completed: bool = False

# --- PLAYER & GAME STATE ---
class Player(BaseModel):
    player_id: str
    secret_token: str # Used for authentication and rejoin
    score: int = 0
    train_cars_remaining: int = 45
    hand: List[TrainCard] = []
    goals: List[GoalCard] = []
    pending_goals: List[GoalCard] = []

class GameState(BaseModel):
    game_id: str
    status: GameStatus = GameStatus.LOBBY
    
    turn_order: List[str] = []
    current_turn_index: int = 0
    cards_drawn_this_turn: int = 0
    final_turns_remaining: int = 0
    
    cities: Dict[str, City]
    board: Dict[str, Route]
    
    train_deck: List[TrainCard] = []
    train_discard: List[TrainCard] = []
    face_up_trains: List[TrainCard] = []
    goal_deck: List[GoalCard] = []
    players: Dict[str, Player]

File: frontend/src/types/game.ts
export enum TrainColor {
  RED = "red",
  BLUE = "blue",
  GREEN = "green",
  YELLOW = "yellow",
  BLACK = "black",
  WHITE = "white",
  ORANGE = "orange",
  PINK = "pink",
  WILD = "wild",
}

export enum RouteColor {
  RED = "red",
  BLUE = "blue",
  GREEN = "green",
  YELLOW = "yellow",
  BLACK = "black",
  WHITE = "white",
  ORANGE = "orange",
  PINK = "pink",
  ANY = "any",
}

export enum GameStatus {
  LOBBY = "lobby",
  SETUP = "setup",
  ACTIVE = "active",
  FINAL_ROUND = "final_round",
  FINISHED = "finished",
}

export interface City {
  id: string;
  name: string;
  x: number;
  y: number;
}

export interface Route {
  id: string;
  node_a: string;
  node_b: string;
  length: number;
  color: RouteColor;
  required_wildcards: number;
  owner_id: string | null;
  sibling_id: string | null;
}

export interface TrainCard {
  id: string;
  color: TrainColor;
}

export interface GoalCard {
  id: string;
  node_a: string;
  node_b: string;
  points: number;
  is_completed: boolean;
}

export interface Player {
  player_id: string;
  score: number;
  train_cars_remaining: number;
  hand: TrainCard[];
  goals: GoalCard[];
  pending_goals: GoalCard[];
}

export interface GameState {
  game_id: string;
  status: GameStatus;
  turn_order: string[];
  current_turn_index: number;
  cards_drawn_this_turn: number;
  final_turns_remaining: number;
  cities: Record<string, City>;
  board: Record<string, Route>;
  face_up_trains: TrainCard[];
  players: Record<string, Player>;
  // We don't expose decks to the client in the same way, 
  // but we might need counts if the backend sends them.
  train_deck_count?: number;
  goal_deck_count?: number;
}
